Connection leakage in CustomRestClient

### Describe the bug
We found that `ConnectionPoolTimeoutException` is thrown from `CustomRestClientImpl` and cause custom health check failure.

```
[INFO ] 2020-10-16 13:10:54,273 org.apache.helix.rest.client.CustomRestClient - Executing request: POST http://localhost:11933/instanceHealthStatus HTTP/1.1, headers:
[Accept: application/json], entity: [Content-Type: application/json; charset=UTF-8,Content-Length: 2,Chunked: false]
[ERROR] 2020-10-16 13:11:54,281 org.apache.helix.rest.client.CustomRestClient - Failed to perform customized health check. Is participant endpoint http://localhost:119
33/instanceHealthStatus available?
org.apache.http.conn.ConnectionPoolTimeoutException: Timeout waiting for connection from pool
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.leaseConnection(PoolingHttpClientConnectionManager.java:314)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager$1.get(PoolingHttpClientConnectionManager.java:280)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:190)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
        at org.apache.helix.rest.client.CustomRestClientImpl.post(CustomRestClientImpl.java:140)
        at org.apache.helix.rest.client.CustomRestClientImpl.getInstanceStoppableCheck(CustomRestClientImpl.java:75)
        at org.apache.helix.rest.server.service.InstanceServiceImpl.performCustomInstanceCheck(InstanceServiceImpl.java:259)
        at org.apache.helix.rest.server.service.InstanceServiceImpl.lambda$null$2(InstanceServiceImpl.java:200)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### To Reproduce
Change code a bit to let `handleResponse` always go to the else branch that does not consume entity and throw `ClientProtocolException`.

### Expected behavior
Should not have `ConnectionPoolTimeoutException`.

### Additional context
If custom health check rest calls return non-200 responses, the entity is not consumed so the connection is still occupied, which causes connection leakage. We need to consume the response entity before throwing an exception for non-200 responses, so the stream or connection is released.

--- a/helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java
+++ b/helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java
@@ -40,7 +40,7 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 class CustomRestClientImpl implements CustomRestClient {
-  private static final Logger LOG = LoggerFactory.getLogger(CustomRestClient.class);
+  private static final Logger LOG = LoggerFactory.getLogger(CustomRestClientImpl.class);
 
   // postfix used to append at the end of base url
   private static final String INSTANCE_HEALTH_STATUS = "/instanceHealthStatus";
@@ -116,6 +116,8 @@ class CustomRestClientImpl implements CustomRestClient {
       LOG.info("Expected HttpResponse statusCode: {}", HttpStatus.SC_OK);
       return jsonConverter.convert(getJsonObject(httpResponse));
     } else {
+      // Ensure entity is fully consumed so stream is closed.
+      EntityUtils.consumeQuietly(httpResponse.getEntity());
       throw new ClientProtocolException("Unexpected response status: " + status + ", reason: "
           + httpResponse.getStatusLine().getReasonPhrase());
     }
@@ -123,18 +125,27 @@ class CustomRestClientImpl implements CustomRestClient {
 
   @VisibleForTesting
   protected HttpResponse post(String url, Map<String, String> payloads) throws IOException {
+    HttpPost postRequest = new HttpPost(url);
     try {
-      HttpPost postRequest = new HttpPost(url);
       postRequest.setHeader("Accept", ACCEPT_CONTENT_TYPE);
       StringEntity entity = new StringEntity(OBJECT_MAPPER.writeValueAsString(payloads),
           ContentType.APPLICATION_JSON);
       postRequest.setEntity(entity);
       LOG.info("Executing request: {}, headers: {}, entity: {}", postRequest.getRequestLine(),
           postRequest.getAllHeaders(), postRequest.getEntity());
-      return _httpClient.execute(postRequest);
+
+      HttpResponse response = _httpClient.execute(postRequest);
+      int status = response.getStatusLine().getStatusCode();
+      if (status != HttpStatus.SC_OK) {
+        LOG.warn("Received non-200 status code: {}, payloads: {}", status, payloads);
+      }
+
+      return response;
     } catch (IOException e) {
       LOG.error("Failed to perform customized health check. Is participant endpoint {} available?",
           url, e);
+      // Release connection to be reused and avoid connection leakage.
+      postRequest.releaseConnection();
       throw e;
     }
   }

